### PR TITLE
966 translate all existing frameworks to fdl

### PR DESCRIPTION
**Pull request by rgarner**
# [Translate ALL existing frameworks to FDL](https://trello.com/c/AxbhXG3x/966-translate-all-existing-frameworks-to-fdl)

## The `rails g framework:fdl`

Because there were a lot, and because it's pretty error-prone work that requires a lot of revisiting if we edit the `.rb` definitions, we wrote a temporary generator to introspect the existing Ruby classes and output field definitions and lookups. It *doesn't*: do dependent field validations/marry up the disparate constant names of lookups to known fields. So there's still a bit of manual work required, but it reduces the possibility of error a fair bit.

## Why we've removed `presence: true` from some `.rb` definitions

Where validators are stricter, like in the case of `ingested_date`, for example, FDL doesn't generate a `presence: true` validator. In order to keep the results of `rake fdl:validations:test` usable, we remove the validators that FDL wouldn't generate from the original definitions.

## Big caveat (no longer applies, we've just loosened all the validations)

*NB* This isn't all existing frameworks. Some aren't here, because we'd need to loosen the validation around their known fields. See [Allow validation of wide known field types as narrower types for validation purposes](https://trello.com/c/AxbhXG3x/966-translate-all-existing-frameworks-to-fdl). Suggest that for those frameworks we either:

1. Implement a language feature that allows us for the purposes of validation to ensure that a wider field can be validated as a narrower type (so a `String` for example could be validated as an `Integer` or a `Decimal` as an `Integer)
2. Get agreement that we don't need to do this, and drop all validation for things like the Digital Marketplace Service ID

### CODE REVIEW

File: lib/active_model/introspector/lookup.rb [hunk 1]
```diff
@@ -0,0 +1,16 @@
+module ActiveModel
```
Comment (james) [inline lib/active_model/introspector/lookup.rb]: So I think if this is part of the FDL generator, then it should be in that namespace, and can be removed when all of that is removed.

If it's going to hang around afterwards, as @rgarner has mentioned in DMs, then I'd like to have a better explanation of what is for, as I'm not entirely following what use it has outside of our FDL generator.

File: lib/active_model/introspector/lookups.rb [hunk 2]
```diff
@@ -17,7 +17,7 @@ def each
 
       private
 
-      ONLY_ALL_CAPS_CONSTS = /^[A-Z_]*$/
+      ONLY_ALL_CAPS_CONSTS = /^[A-Z0-9_]*$/.freeze
```
Comment (james) [inline lib/active_model/introspector/lookups.rb]: Should this have a test to catch this in the future?

File: app/models/framework/definition/RM3786.fdl [hunk 3]
```diff
@@ -0,0 +1,148 @@
+Framework RM3786 {
```
Comment (james) [inline app/models/framework/definition/RM3786.fdl]: I'm assuming this framework passes the validation task? You mentioned that they had in previous commit messages, so just making sure.

File: app/models/framework/definition/RM3710.fdl [hunk 4]
```diff
@@ -0,0 +1,60 @@
+Framework RM3710 {
```
Comment (james) [inline app/models/framework/definition/RM3710.fdl]: Same here. Do I just assume that you've validated them all?

File: app/models/framework/definition/RM3786.fdl (hunk 3 above)
Comment (rgarner) [reply app/models/framework/definition/RM3786.fdl]: Yes, I put all the framework names in a local `all_frameworks.txt` and ran

```
for FRAMEWORK in $(cat all_frameworks.txt); do echo $FRAMEWORK && rake fdl:validations:test[$FRAMEWORK,1000]; done
```

So they all pass for a sample of 1000 rows.  More than that means an overnight run :)

File: app/models/framework/definition/RM3710.fdl (hunk 4 above)
Comment (rgarner) [reply app/models/framework/definition/RM3710.fdl]: Yes. But only against a rotating sample of 1000.

File: lib/active_model/introspector/lookups.rb (hunk 2 above)
Comment (rgarner) [reply lib/active_model/introspector/lookups.rb]: No. It's going away.

File: lib/active_model/introspector/lookup.rb (hunk 1 above)
Comment (rgarner) [reply lib/active_model/introspector/lookup.rb]: The main use it *would* have is generating a visual Ruby representation of our FDL anonymous classes for debugging purposes. By which I mean this namespace; this file will go away.

Comment (rgarner) [reply lib/active_model/introspector/lookup.rb]: The main problem you'll have in support is introspecting those anonymous classes; ActiveModel by itself doesn't make it easy to see metadata about attributes (the thing that has the `AttributeSet` with all its types and validators on is in an `@attributes` field with no accessor).

Comment (rgarner) [reply lib/active_model/introspector/lookup.rb]: The introspector can help here in a Rails console, but you probably also want something like `rake fdl:visualise[RM3787]` to print what the actual Ruby would be. You may of course survive without it!

File: app/models/framework/definition/ast/field/options.rb [hunk 5]
```diff
@@ -39,7 +39,7 @@ def no_presence_required?
             # YesNo fields don't require an accompanying +presence: true+
```
Comment (james) [inline app/models/framework/definition/ast/field/options.rb]: Does this mean this comment is now out of date?

File: app/models/framework/definition/RM1043_5.rb [hunk 6]
```diff
@@ -85,9 +85,9 @@ class Order < EntryData
         field 'Customer Organisation Name', :string, exports_to: 'CustomerName', presence: true
         field 'Customer Unique Reference Number (URN)', :integer, exports_to: 'CustomerURN', urn: true
         field 'Lot Number', :string, exports_to: 'LotNumber', presence: true, lot_in_agreement: true
-        field 'Contract Start Date', :string, exports_to: 'ContractStartDate', ingested_date: true, presence: true
```
Comment (james) [inline app/models/framework/definition/RM1043_5.rb]: Is this in the wrong commit?

File: app/models/framework/definition/RM3787.fdl [hunk 7]
```diff
@@ -36,7 +36,7 @@ Framework RM3787 {
     CustomerName from 'Customer Organisation Name'
     CustomerPostcode from 'Customer Post Code'
     CustomerReferenceNumber from 'Matter Name'
-    ProductDescription from 'Matter Description'
+    String Additional7 from 'Matter Description'
```
Comment (james) [inline app/models/framework/definition/RM3787.fdl]: Should we update the ruby definition to reflect this too?

**Comment by james** [comment]
First off, amazing work! Thanks for doing this.

There are a few remaining comments I've just added, but once those have been resolved I'd like to run your validate all FDL definitions for all frameworks script.

I have also put together a table of all the changes (apart from removing redundant presence validators) that are made to the ruby definitions.

Framework | Field | Change
-- | -- | --
RM1043_5 | Opportunity ID | Remove ingested_numericality and min length 4
RM1070 | Quantity | Allow decimals, not just integers
RM3772 | Customer Order/Contract Value | Add validation of ingested_numericality
RM3786 | Contract Start Date | Add ingested_date: true, allow_nil: true
RM1557_10 | Digital Marketplace Service ID | Remove ingested_numericality
RM1557ix | Digital Marketplace Service ID | Remove ingested_numericality
RM1557vii | Digital Marketplace Service ID | Remove ingested_numericality
RM1557viii | Digital Marketplace Service ID | Remove ingested_numericality

I wanted to put this together so I could run it past Timur and double check that we're OK with all of them.

File: app/models/framework/definition/ast/field/options.rb (hunk 5 above)
Comment (rgarner) [reply app/models/framework/definition/ast/field/options.rb]: It does.

File: app/models/framework/definition/RM3787.fdl (hunk 7 above)
Comment (rgarner) [reply app/models/framework/definition/RM3787.fdl]: Yes.

**Comment by rgarner** [comment]
@james thank you greatly for the table, this was something I wanted to do and didn't get time.

File: app/models/framework/definition/RM1043_5.rb (hunk 6 above)
Comment (rgarner) [reply app/models/framework/definition/RM1043_5.rb]: Yes.

File: app/models/framework/definition/RM3787.fdl (hunk 7 above)
Comment (rgarner) [reply app/models/framework/definition/RM3787.fdl]: Well, maybe. If we do it now it takes effect on next deploy. If we don't it takes effect when the FDL supersedes this. I'm guessing you'd prefer the first?

Comment (james) [reply app/models/framework/definition/RM3787.fdl]: Yes, I'd like to have basically no feature changes when we switch from one to that other, and I can add it to the table below and give to Timur.

**Pull request merged**